hw02-DIYArrayList/src/main/java/com/vasiliyzhigalov/DIYArrayList.java

```diff
@@ -0,0 +1,251 @@
+        for (int i = pointer; i >= validIndex; i--)
+            array[i + 1] = array[i];
+        pointer++;
+    }
+
+    @Override
+    public E remove(int index) {
+        int validIndex = checkIndex(index, pointer);
+        for (int i = validIndex; i < pointer; i++)
+            array[i] = array[i + 1];
+        pointer--;
+        return (E) array;
+    }
+
+    @Override
+    public int indexOf(Object o) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int lastIndexOf(Object o) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+
+    public ListIterator<E> listIterator() {
+        return new ListIterator<E>() {
+            private int currentIndex = 0;
+            private int lastIndex = -1;
+
+            @Override
+            public boolean hasNext() {
+                return currentIndex < pointer && array[currentIndex] != null;
+            }
+
+            @Override
+            public E next() {
+                E result;
+                if (!hasNext()) {
+                    throw new NoSuchElementException();
+                } else {
+                    lastIndex = currentIndex;
+                    result = get(currentIndex);
+                    currentIndex++;
+                }
+                return result;
+            }
+
+            @Override
+            public boolean hasPrevious() {
+                return currentIndex > 0;
+            }
+
+            @Override
+            public E previous() {
+                if (!hasPrevious()) {
+                    throw new NoSuchElementException();
+                } else {
+                    lastIndex = currentIndex;
+                    return (E) array[--currentIndex];
+                }
+            }
+
+            @Override
+            public int nextIndex() {
+                return currentIndex;
+            }
+
+            @Override
+            public int previousIndex() {
+                return currentIndex - 1;
+            }
+
+            @Override
+            public void remove() {
+
+            }
+
+            @Override
+            public void set(E e) {
+                array[lastIndex] = e;
+            }
+
+            @Override
+            public void add(E e) {
+            }
+        };
+    }
+
+    @Override
+    public ListIterator<E> listIterator(int index) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public List<E> subList(int fromIndex, int toIndex) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public String toString() {
+        Object[] result = new Object[pointer];
+        System.arraycopy(array, 0, result, 0, pointer);
+        return Arrays.toString(result);
+    }
+
+    private int checkIndex(int index, int length) {
+        if (index >= 0 && index < length)
+            return index;
+        else {
+            throw new IndexOutOfBoundsException();
```

Comment:
```suggestion
            throw new IndexOutOfBoundsException(index);
```

Чтобы по тексту исключения быстрее понять, в чем была проблема.